src/sinks/util/retries.rs

```diff
@@ -22,13 +22,18 @@ pub enum RetryAction {
 pub trait RetryLogic: Clone + Send + Sync + 'static {
     type Error: std::error::Error + Send + Sync + 'static;
     type Response;
+    type Request;
 
     fn is_retriable_error(&self, error: &Self::Error) -> bool;
 
     fn should_retry_response(&self, _response: &Self::Response) -> RetryAction {
         // Treat the default as the request is successful
         RetryAction::Successful
     }
+
+    fn clone_request(&self, _request: &Self::Request) -> Option<Self::Request> {
+        None
```

Comment:
Hm, I think it's worth changing the default implementation to `Some(request.clone())`, but I think we still need return `Option` because it's more flexible. At least this will allow to remove `or_else`.

`clone_request` because what we return is not actually request which will be used right now, it's request with will be used on next failed response.